### ISSUE TITLE
Discrepancy in IPOPT convergence behaviour with/without derivative checker and BFGS

### ISSUE THREAD

**Issue by johnTrumpet** [coin-or/Ipopt#435]
I am experiencing different convergence behaviors of IPOPT on a few problems in the `MINLPLIB` benchmark, like `fuel.mod`. I set the following options from C++:
```
nlpOptsPtr->SetIntOption( "print_level", 5 );
nlpOptsPtr->SetIntOption( "max_iter", 200 );
nlpOptsPtr->SetStrOption( "hessian_approximation", "limited-memory" );
nlpOptsPtr->SetStrOption( "limited_memory_update_type", "bfgs" );
nlpOptsPtr->SetStrOption( "derivative_test", "none" );
```

I have built a custom modeler to parse `.mod` file and compute derivatives via AD. When I use the following options on `fuel.mod`, IPOPT goes to maxit (200). If I set `first-order` in the derivative checker I get convergence to a locally optimal solution in 92 iterations. **I initially thought that this likely to be a bug in my modeler, but after a thorough check I reached the conclusion that it is not: all symbolic expressions provide the same values in both cases up to 2 iterations, at the 3rd iteration the primal iterate differs between the run with derivative checker and the run without derivative checker. This seems to indicate a different behavior within IPOPT.** 
Could you please clarify ? Is the derivative checker impacting the IPOPT algorithm in any way or should I check further in my modeler ?

Note that I am using ipopt 3.12.13 built with MUMPS. Valgrind does not show any issue. The issue does not occur on all problems.

**Comment by johnTrumpet** [comment]
Is the gradient computed during the derivative check before starting iterations used in the limited memory updates if the check did not return any error ?

**Comment by johnTrumpet** [comment]
Seems this is potentially related to the limited-memory approximation. When setting `limited_memory_max_history` to `0` or `1` with derivative checker enabled, IPOPT also goes to maxit. So it seems that the point used by the derivative checker is also used in the limited-memory history. Could you please confirm ? If this is the case, the behavior is pretty confusing. Is this mentioned anywhere in the doc ?

**Comment by svigerske** [comment]
I don't see where this would be happening.
As far as I see from the derivate checking code (https://github.com/coin-or/Ipopt/blob/master/src/Interfaces/IpTNLPAdapter.cpp#L2772), it only communicates with your TNLP, but does not store the obtained gradients somewhere in Ipopt.

When I run hs071_cpp with
```
print_level 6
hessian_approximation limited-memory
derivative_test first-order
```
then the messages
```
Starting derivative checker for first derivatives.


No errors detected by derivative checker.
```
come before
```
**************************************************
*** Update HessianMatrix for Iteration 0:
**************************************************

Limited-Memory approximation started; store data at current iterate.
```
(https://github.com/coin-or/Ipopt/blob/master/src/Algorithm/IpLimMemQuasiNewtonUpdater.cpp#L144)

So it seems to me that the Hessian approximation is initialized after the derivate test is done and that there is no use of gradients from the derivate test for the Hessian approximation.

You might want to confirm this with your code by also running with a higher `print_level` and see where things start to differ.
Also check that your TNLP implementation does not accidentally reuse no-longer-valid information from previous evaluation calls (check the use of `new_x`).

To get more output, configure Ipopt with `--with_ipopt_verbosity=1` and change `dbg_verbosity` to 1 in `IpLimMemQuasiNewtonUpdater.cpp`. Running Ipopt should then create a file debug.out which includes printouts of the vectors and matrices compued by the Hessian updater.

**Comment by johnTrumpet** [comment]
Thanks a lot for clarifying Stefan. Might still be an issue with `new_x` in my code. Will keep you posted.

**Comment by svigerske** [comment]
Closing as I don't see what else could be done from my side here at the moment.

**Comment by svigerske** [comment]
I ran into a problem with my code where enabling the derivative-checker changed the solve of Ipopt. It turned out that the problem was that with the derivative-checker enabled, Ipopt queries the starting point twice from the TNLP instead of only one time. In my case, I generated a random starting point, so I reported a different point for the second call of `TNLP::get_starting_point()`, which is the point that Ipopt actually uses to initialize its initial iterate.
Thought I share this here as it has same symptoms as reported in this issue.
